### ISSUE TITLE
BUNDLE: Ref Re-Mapping Broken 

### ISSUE THREAD

**Issue by hipstersmoothie** [BigstickCarpet/json-schema-ref-parser#67]
It seems like when i have a sufficiently complex schema ref-re-mapping breaks. 

I am getting the error (in VSCode, but the error is in the schema):

```
$ref '/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/affirmativeAction/oneOf/1/properties/metaData' in 'http://localhost:8000/merged-schema.json' can not be resolved.
```

this points to 

```json
    "gatewayActions": {
        "type": "object",
        "properties": {
            "binding": {
                "type": "string"
            },
            "affirmativeAction": {
                "$ref": "#/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction"
            },
            "negativeAction": {
                "$ref": "#/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction"
            },
            "prevAction": {
                "title": "Action Choice",
                "oneOf": [
                    {
                        "title": "Action Asset",
                        "type": "object",
                        "required": [
                            "asset"
                        ],
                        "properties": {
                            "asset": {
                                "$ref": "#/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/8"
                            }
                        }
                    },
                    {
                        "title": "Action Object",
                        "description": "(DEPRECATED): Use an action asset instead.",
                        "type": "object",
                        "not": {
                            "required": [
                                "asset"
                            ]
                        },
                        "properties": {
                            "id": {
                                "type": "string"
                            },
                            "label": {
                                "$ref": "#/properties/views/items/oneOf/5/allOf/1/properties/coveredForms/properties/label"
                            },
                            "metaData": {
                                "id": "metaData",
                                "type": "object",
                                "additionalProperties": true,
                                "properties": {
                                    "role": {
                                        "type": "string",
                                        "description": "A role assigned that gives it greater semantic meaning."
                                    }
                                }
                            },
                            "value": {
                                "type": "string"
                            },
                            "applicability": {
                                "type": "string"
                            },
                            "accessibility": {
                                "type": "string"
                            }
                        }
                    }
                ]
            },
            "actions": {
                "maximum": 0
            }
        }
    }
```

Instead of a object, I get reference to a sibling. I believe that all references to:

```
/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/affirmativeAction/oneOf/1/properties/metaData
```
should be 

```
/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction/oneOf/1/properties/metaData
```

Full Schema, Code, and Error:
https://paste.ee/p/weaNo

**Comment by hipstersmoothie** [comment]
I think that maybe whats happening is you have one schema that has sub schemas being cached first.

a.json
```json
{
    "$id": "a",
     "properties": {
        "metaData": {
            "properties": {
                "id": {
                    "type": "string"
                }
            }
        },
    }
}
```

b.son
```json
{
    "$id": "b",
    "properties": {
        "metaData": {
            "$ref": "a#/properties/metaData"
        },
    }
}
```

c.son
```json
{
    "$id": "c",
    "properties": {
        "someProp": {
            "$ref": "a"
        },
    }
}
```


1. b's metaData ref points to 'a#/properties/metaData'
2. a gets cached at some point for c -> c's someProp points to a which points to another object
3. b's metadata pointer is now invalid

**Comment by hipstersmoothie** [comment]
Looking at my debug output is see this:

```
Re-mapping $ref pointer "#/definitions/actions" at #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
    new value: [object Object]
Re-mapping $ref pointer "#/definitions/actions" at #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/affirmativeAction
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "#/definitions/actions" at #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/negativeAction
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "#/definitions/actions" at #/properties/views/items/oneOf/0/allOf/0/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/7/allOf/1/properties/landingTable/properties/asset/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/10/allOf/1/survey/properties/asset/allOf/1/properties/surveyItems/items/properties/prevAction
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/10/allOf/1/survey/properties/asset/allOf/1/properties/surveyItems/items/properties/nextAction
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/33/definitions/question_item/oneOf/0
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/28/allOf/1/properties/template
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/36/allOf/1/properties/template
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/41/allOf/1/properties/template
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/18/allOf/1/properties/template
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/14/allOf/1/properties/template
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/18/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/5/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/42/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/11/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/28/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/13/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/25/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/14/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/24/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/36/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/41/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/27/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/13/allOf/1/properties/template/oneOf/0
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/11/definitions/value_item/properties/asset/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/18/allOf/1/properties/value/properties/asset/allOf/1/properties/actions/items
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/32/allOf/1/properties/negativeAction
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
Re-mapping $ref pointer "def#/definitions/actions" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/32/allOf/1/properties/affirmativeAction
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/prevAction
```

and later

```
Re-mapping $ref pointer "metaData" at #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/affirmativeAction/oneOf/1/properties/metaData
    new value: [object Object]
Re-mapping $ref pointer "metaData" at #/properties/views/items/oneOf/10/allOf/1/survey/properties/asset/allOf/0/properties/metaData
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/affirmativeAction/oneOf/1/properties/metaData
Re-mapping $ref pointer "metaData" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/4/definitions/value_item/allOf/1/properties/metaData
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/affirmativeAction/oneOf/1/properties/metaData
Re-mapping $ref pointer "metaData" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/0/allOf/1/properties/modifiers/items/anyOf/0/properties/metaData/allOf/0
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/affirmativeAction/oneOf/1/properties/metaData
Re-mapping $ref pointer "metaData" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/15/allOf/1/properties/modifiers/items/anyOf/2/properties/metaData/allOf/0
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/affirmativeAction/oneOf/1/properties/metaData
Re-mapping $ref pointer "metaData" at #/properties/views/items/oneOf/1/allOf/1/properties/fields/definitions/asset/oneOf/0/allOf/1/properties/modifiers/items/anyOf/7/properties/metaData/allOf/0
    new value: #/properties/views/items/oneOf/5/allOf/1/properties/gatewayActions/properties/affirmativeAction/oneOf/1/properties/metaData
```

so the more specific remapping (action/metaData) is happening after the less specific one (just action). this results in the first more specific one to get overwritten and any reference to it is a bug.

**Comment by hipstersmoothie** [comment]
I have investigated further and here is the minimum code i need to reproduce the bug:

https://gist.github.com/hipstersmoothie/21906ec9d2ca9525b066224d77ab8eff

**Comment by BigstickCarpet** [comment]
This issue is fixed in [version 5.1.0](https://github.com/BigstickCarpet/json-schema-ref-parser/blob/master/CHANGELOG.md#v510-2018-07-11)
